include/oneapi/tbb/detail/_flow_graph_impl.h

```diff
@@ -165,22 +165,44 @@ class graph_task_with_message_waiters : public graph_task {
         }
     }
 
+    graph_task_with_message_waiters(graph& g, d1::small_object_allocator& allocator,
+                                    node_priority_t node_priority,
+                                    std::forward_list<d1::wait_context_vertex*>&& msg_waiters)
+        : graph_task(g, allocator, node_priority)
+        , my_msg_wait_context_vertices(std::move(msg_waiters))
+    {
+    }
+
     graph_task_with_message_waiters(graph& g, d1::small_object_allocator& allocator,
                                     const std::forward_list<d1::wait_context_vertex*>& msg_waiters)
         : graph_task_with_message_waiters(g, allocator, no_priority, msg_waiters) {}
 
+    graph_task_with_message_waiters(graph& g, d1::small_object_allocator& allocator,
+                                    std::forward_list<d1::wait_context_vertex*>&& msg_waiters)
+        : graph_task_with_message_waiters(g, allocator, no_priority, std::move(msg_waiters)) {}
+
     const std::forward_list<d1::wait_context_vertex*> get_msg_wait_context_vertices() const {
         return my_msg_wait_context_vertices;
     }
 
 protected:
     template <typename DerivedType>
     void finalize(const d1::execution_data& ed) {
+        auto wait_context_vertices = std::move(my_msg_wait_context_vertices);
         auto msg_reference_vertices = std::move(my_msg_reference_vertices);
         graph_task::finalize<DerivedType>(ed);
 
-        for (auto& msg_waiter : msg_reference_vertices) {
-            msg_waiter->release(1);
+        // If there is no thread reference vertices associated with the task
+        // then this task was created by transferring the ownership from other metainfo
+        // instance (e.g. while taking from the buffer)
+        if (msg_reference_vertices.empty()) {
+            for (auto& msg_waiter : wait_context_vertices) {
+                msg_waiter->release(1);
+            }
+        } else {
+            for (auto& msg_waiter : msg_reference_vertices) {
+                msg_waiter->release(1);
+            }
```

Comment:
Does it make sense to move this check to object initialization? I'm not sure if this is possible:
```
std::forward_list<wait_tree_vertex_interface*> vertices = msg_reference_vertices.empty() ? std::move(my_msg_wait_context_vertices) : std::move(my_msg_reference_vertices);
```

Reply:
It is possible, but not as straightforward as you proposed because `my_msg_wait_context_vertices` and `my_msg_reference_vertices` has different types so you can't just move one to another, only to do a per-element move, but I believe it it not what we want here.

Reply:
So container with base class type won't work?